.buildkite/e2e/pipeline-gen/main.go

```diff
@@ -0,0 +1,390 @@
+type TestsSuiteRun struct {
+	Name     string
+	SlugName string
+	Env      Env
+	Dind     bool
+	Cleanup  bool
+}
+
+func newTestsSuite(groupLabel string, fixed Env, mixedLen int, mixed Env) (TestsSuiteRun, error) {
+	resolveShortcuts(fixed)
+	resolveShortcuts(mixed)
+
+	// find k8s provider
+	provider, ok := fixed[EnvVarProvider]
+	if !ok {
+		provider, ok = mixed[EnvVarProvider]
+		if !ok {
+			return TestsSuiteRun{}, fmt.Errorf("%s not defined", EnvVarProvider)
+		}
+	}
+
+	name := getName(groupLabel, provider, mixedLen, mixed)
+	slugName := getSlugName(name)
+	env, err := commonTestEnv(name, slugName)
+	if err != nil {
+		return TestsSuiteRun{}, err
+	}
+	t := TestsSuiteRun{
+		Name:     name,
+		SlugName: slugName,
+		Dind:     slices.Contains(providersInDocker, provider),
+		Cleanup:  !slices.Contains(providersNoCleanup, provider),
+		Env:      env,
+	}
+
+	// merge fixed and mixed vars
+	for k, v := range fixed {
+		t.Env[k] = v
+	}
+	for k, v := range mixed {
+		t.Env[k] = v
+	}
+
+	return t, nil
+}
+
+func resolveShortcuts(e Env) {
+	for k, v := range e {
+		for short, long := range shortcuts {
+			if k == short {
+				e[long] = v
+				delete(e, short)
+			}
+		}
+	}
+}
+
+func getName(groupLabel, provider string, mixedLen int, mixed Env) string {
+	name := groupLabel
+	if name == "" {
+		name = provider
+	}
+
+	if mixedLen < 2 {
+		return name
+	}
+
+	// try use the two first env var values as suffix if more than one var in the mixed vars
+
+	suffixes := make([]string, 0)
+	i := 0
+	for _, val := range mixed {
+		suffix := val
+		// do not repeat name
+		if suffix == name {
+			continue
+		}
+		// extract semver (e.g.: kind node image)
+		match := semverRE.FindStringSubmatch(suffix)
+		if len(match) > 0 {
+			suffix = match[0]
+		}
+		suffixes = append(suffixes, suffix)
+		i++
+		if i == 2 {
+			break
+		}
+	}
+	sort.Strings(suffixes)
+
+	if len(suffixes) == 0 {
+		return name
+	}
+
+	return fmt.Sprintf("%s-%s", name, strings.Join(suffixes, "-"))
+
+}
+
+func getSlugName(name string) string {
+	// sanitize
+	name = strings.ReplaceAll(name, ".", "-")
+	name = strings.ReplaceAll(name, ":", "-")
+	name = strings.ReplaceAll(name, "/", "-")
+	name = strings.ToLower(name)
+
+	// truncate
+	if len(name) > 16 {
+		name = name[:16]
+	}
+
+	// random id
+	id := make([]rune, 4)
+	for i := range id {
+		id[i] = chars[rand.Intn(len(chars))]
+	}
+
+	return fmt.Sprintf("%s-%s", name, string(id))
+}
+
+func commonTestEnv(name string, slugName string) (map[string]string, error) {
+	buildNumber, ok := os.LookupEnv(EnvVarBuildkiteBuildNumber)
+	if !ok {
+		buildNumber = "0"
+	}
+
+	operatorImage, err := getMetadata("operator-image")
+	if err != nil {
+		return nil, err
+	}
+	e2eImage, err := getMetadata("e2e-image")
+	if err != nil {
+		return nil, err
+	}
+
+	operatorImageSuffix := os.Getenv(EnvVarBuildLicensePubkey)
+	if operatorImageSuffix != "" {
+		operatorImage = fmt.Sprintf("%s-%s", operatorImage, operatorImageSuffix)
```

Comment:
This needs at least a comment to explain why we are appending the value of a variable that is judging by its name about license to the operator image. 

I am assuming the values are either `dev` or `` but still I am confused.

Reply:
It's not useful today but the values could be anything.

I can add a comment but I admit I am not very satisfied with my refactoring of the operator image build to test snapshots. I think the main issue is that I use the variable `BUILD_LICENSE_PUBKEY` for 2 things: to discover the pub key to use but also implicitly to suffix the image, which is confusing I agree.

```js
pipeline-build
  if nightly-build
    additional build-operator-image with BUILD_LICENSE_PUBKEY
      -> use BUILD_LICENSE_PUBKEY to get a special pubkey embedded in the operator (cmd/license-initializer)
      -> push image suffixed by -$BUILD_LICENSE_PUBKEY

pipeline-e2e
  pipeline-gen
    if BUILD_LICENSE_PUBKEY
      -> use image suffixed by -$BUILD_LICENSE_PUBKEY
```

We could add a more explicit OPERATOR_IMAGE_SUFFIX which will be a repeat of BUILD_LICENSE_PUBKEY during build and will be used for testing instead of BUILD_LICENSE_PUBKEY?

```js
pipeline-build
  if nightly-build
    additional build-operator-image with BUILD_LICENSE_PUBKEY and OPERATOR_IMAGE_SUFFIX
      -> use BUILD_LICENSE_PUBKEY to get a special pubkey embedded in the operator (cmd/license-initializer)
      -> push image suffixed by -$OPERATOR_IMAGE_SUFFIX

pipeline-e2e
  pipeline-gen
    if OPERATOR_IMAGE_SUFFIX
      -> use image suffixed by -$OPERATOR_IMAGE_SUFFIX
```

This makes me realize that if we want to test snapshot stack version from a PR, the `-dev`operator image is currently not built.

Reply:
Do you plan to address this in this PR or a subsequent one? Otherwise I think this PR looks 👍 to me.

Reply:
I opened https://github.com/elastic/cloud-on-k8s/issues/6567 to address this in another PR.